vignettes/opencl.Rmd

```diff
@@ -0,0 +1,143 @@
+---
+title: "Run Stan on the GPU with OpenCL"
+author: "Rok Češnovar and Jonah Gabry"
+output: 
+  rmarkdown::html_vignette:
+    toc: true
+    toc_depth: 4
+params:
+  EVAL: !r identical(Sys.getenv("CMDSTANR_OPENCL_TESTS"), "true")
+vignette: >
+  %\VignetteIndexEntry{Run Stan on the GPU with OpenCL}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction
+
+This vignette demonstrates how to use the OpenCL capabilites of CmdStan with
+CmdStanR. The vignette requires CmdStan 2.26.1 or newer.
+
+As of version 2.26.1, users can expect speedups with OpenCL when using vectorized
+probability distribution/mass functions (functions with the `_lpdf` or `_lpmf`
+suffix). You can expect speedups when the input variables contain 20.000 or more elements.
+
+The actual speedup for a model will depend on whether the `lpdf/lpmf` functions
+are the bottlenecks of the model and the `lpdf/lpmf` function used. 
+The more computationally complex the function is, the larger the expected speedup.
+The biggest speedups are expected when using the GLM functions.
+
+Use [profiling](../profiling.Rmd) in order to establish the bottlenecks in your model.
+
+## OpenCL runtime
+
+OpenCL is supported on most modern CPUs and GPUs. In order to use
+OpenCL in CmdStanR, an OpenCL runtime for the target device must be installed.
+A guide for the most common devices is available [here](https://mc-stan.org/docs/2_26/cmdstan-guide/parallelization.html#opencl).
+
+## Compiling a model with OpenCL
+
+By default, models in CmdStanR are compiled without OpenCL support. Once OpenCL
+support is enabled, a CmdStan model will make use of OpenCL if the functions
+in the model support it. Technically no changes to a model are required to
+support OpenCL since the choice of using OpenCL is handled by the compiler,
+but it can still be useful to rewrite a model to be more OpenCL-friendly by
+using vectorization as much as possible.
+
+Consider a simple logistic regression with parameters `alpha` and `beta`,
+covariates `X`, and outcome `y`.
+
+```stan
+data {
+  int<lower=1> k;
+  int<lower=0> n;
+  matrix[n, k] X;
+  int y[n];
+}
+parameters {
+  vector[k] beta;
+  real alpha;
+}
+model {
+  target += std_normal_lpdf(beta);
+  target += std_normal_lpdf(alpha);
+  target += bernoulli_logit_glm_lpmf(y | X, alpha, beta);
+}
+```
+
+Some fake data will be useful to run this model:
+
+```{r fake-data, message=FALSE, results='hide'}
+library(cmdstanr)
+
+# Generate some fake data
+n <- 200000
```

Comment:
Above you mention 20000 but here you use 200000. Was the intention to use the same number or did you add a zero on purpose? If the latter then maybe use a slightly different number that doesn't look so similar to the first one?

Reply:
20k is meant as the N where CPU is faster than a typical GPU. I used 200k here cause those are N values where one would more likely use a GPU for sampling. I could go with larger `n`/`k`, but didnt want to run longer runs.

Reply:
That makes sense. It's just somewhat confusing because 20000 and 200000 look so similar when reading fast. Maybe use 250000 or any other number that looks a little more different from 20000?

Reply:
Oh,  makes sense, will use 250k

Reply:
Updated now.